pkg/server/policy/policy.go

```diff
@@ -0,0 +1,190 @@
+	FullMethod string `json:"full_method"`
+
+	// Req represents data received from the request body. It MUST be a
+	// protobuf request object with fields that are serializable as JSON,
+	// since they will be used in policy definitions.
+	Req interface{} `json:"req"`
+}
+
+type Result struct {
+	Allow             bool `json:"allow"`
+	AllowIfAdmin      bool `json:"allow_if_admin"`
+	AllowIfLocal      bool `json:"allow_if_local"`
+	AllowIfDownstream bool `json:"allow_if_downstream"`
+	AllowIfAgent      bool `json:"allow_if_agent"`
+}
+
+// NewEngineOrDefault returns a new policy engine. Or if no
+// config is provided, provides the default policy
+func NewEngineOrDefault(cfg *EngineConfig) (*Engine, error) {
+	if cfg == nil || cfg.FileProvider == nil {
+		return DefaultAuthPolicy()
+	}
+	return NewEngine(cfg)
+}
+
+// NewEngine returns a new policy engine. Or nil if no
+// config is provided.
+func NewEngine(cfg *EngineConfig) (*Engine, error) {
+	if cfg == nil || cfg.FileProvider == nil {
+		return nil, errors.New("policy engine config provided is invalid")
+	}
+
+	module, err := os.ReadFile(cfg.FileProvider.PolicyPath)
+	if err != nil {
+		return nil, err
+	}
+	storefile, err := os.Open(cfg.FileProvider.PermissionsPath)
+	if err != nil {
+		return nil, err
+	}
+	defer storefile.Close()
+	store := inmem.NewFromReader(storefile)
+
+	return NewEngineFromRego(string(module), store)
+}
+
+// DefaultAuthPolicy returns the default policy engine
+func DefaultAuthPolicy() (*Engine, error) {
+	var json map[string]interface{}
+	err := util.UnmarshalJSON(defaultPermissionsData, &json)
+	if err != nil {
+		return nil, err
+	}
+	store := inmem.NewFromObject(json)
+
+	return NewEngineFromRego(defaultPolicyRego, store)
+}
+
+// NewEngineFromRego is a helper to create the Engine object
+func NewEngineFromRego(regoPolicy string, dataStore storage.Store) (*Engine, error) {
+	rego := rego.New(
+		rego.Query("data.spire.result"),
+		rego.Package(`spire`),
+		rego.Module("spire.rego", regoPolicy),
+		rego.Store(dataStore),
+	)
+	pr, err := rego.PartialResult(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	e := &Engine{
+		rego: pr,
+	}
+
+	// Test policy with some simple calls to ensure that the
+	// policy can be evaluated properly.
+	if err = e.testPolicy(); err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+func (e *Engine) testPolicy() error {
+	ctx := context.Background()
+	for _, i := range sampleInputs {
+		var inp Input
+		if err := json.Unmarshal([]byte(i), &inp); err != nil {
+			return err
+		}
+
+		if _, err := e.Eval(ctx, inp); err != nil {
+			return fmt.Errorf("policy is misconfigured: %w", err)
+		}
+	}
+	return nil
+}
+
+// Eval determines whether access should be allowed on a resource.
+func (e *Engine) Eval(ctx context.Context, input Input) (result Result, err error) {
+	rs, err := e.rego.Rego(rego.Input(input)).Eval(ctx)
+	if err != nil {
+		return result, err
+	}
+
+	// TODO(tjulian): figure this out
```

Comment:
Heh, my guess would be that the `Eval()` call returns a slice of [rego.Result](https://pkg.go.dev/github.com/open-policy-agent/opa/rego#Result) objects, each of which contains a slice of ExpressionValues.

So in this implementation, we're only looking at the **first** result and the **first** expression value within that result, which means we're potentially throwing out extra information that could be useful for policy evaluation or logging purposes. Maybe there could be "deny" style rules to be grabbed, or different reasons for "allow", I can't remember exactly.

So it could be worth looking into what the rest of those result/expressions could contain, or leave that as a (more detailed) TODO for later

Reply:
Ah i see, since then we've introduced a contract between the policy and the authorization component, specifying specific variables that must be set, and only those variables are queried. These are listed [here](https://github.com/lumjjb/spire/blob/policy_opa/doc/authorization_policy_engine.md#rego-policy). I think this addresses mostly the concerns from this TODO, and like you said we can add on to this list later on.